contracts/examples/VCPayment-v2.sol

```diff
@@ -0,0 +1,228 @@
+
+    /**
+     * @dev mapping of issuerAddress - balance
+     */
+    mapping(address issuerAddress => uint256 balance) public issuerAddressBalance;
+
+    /**
+     * @dev owner balance
+     */
+    uint256 ownerBalance;
+
+    /**
+     * @dev list of paymentDataId - keccak256(abi.encode(issuerId, schemaHash))
+     */
+    bytes32[] internal paymentDataIds;
+
+    /**
+     * @dev Version of contract
+     */
+    string public constant VERSION = '2.0.0';
+
+    /**
+     * @dev Version of contract
+     */
+    uint256 public ownerPartPercent;
+
+    event Payment(
+        uint256 indexed issuerId,
+        string paymentId,
+        uint256 indexed schemaHash,
+        uint256 timestamp
+    );
+
+    error InvalidOwnerPartPercent(string message);
+    error InvalidWithdrawAddress(string message);
+    error PaymentError(string message);
+    error WithdrawError(string message);
+    error OwnerOrIssuerError(string message);
+
+    /**
+     * @dev Owner or issuer modifier
+     */
+    modifier ownerOrIssuer(uint256 issuerId, uint256 schemaHash) {
+        address issuerAddress = paymentData[keccak256(abi.encode(issuerId, schemaHash))]
+            .withdrawAddress;
+        if (issuerAddress != _msgSender() && owner() != _msgSender()) {
+            revert OwnerOrIssuerError('Only issuer or owner can call this function');
+        }
+        _;
+    }
+
+    constructor(uint256 _ownerPartPercent) Ownable(_msgSender()) {
+        if (ownerPartPercent > 0 && ownerPartPercent < 100) {
+            revert InvalidOwnerPartPercent('Invalid owner part percent');
+        }
+        ownerPartPercent = _ownerPartPercent;
+    }
+
+    function updateOwnerPartPercent(uint256 _ownerPartPercent) public onlyOwner {
+        if (ownerPartPercent > 0 && ownerPartPercent < 100) {
+            revert InvalidOwnerPartPercent('Invalid owner part percent');
+        }
+        ownerPartPercent = _ownerPartPercent;
+    }
+
+    function setPaymentValue(
+        uint256 issuerId,
+        uint256 schemaHash,
+        uint256 value,
+        address withdrawAddress
+    ) public onlyOwner {
+        if (withdrawAddress == address(0)) {
+            revert InvalidWithdrawAddress('Invalid withdraw address');
+        }
+        PaymentData memory newPaymentData = PaymentData(
+            issuerId,
+            schemaHash,
+            value,
+            withdrawAddress,
+            0
+        );
+        paymentDataIds.push(keccak256(abi.encode(issuerId, schemaHash)));
+        _setPaymentData(issuerId, schemaHash, newPaymentData);
+    }
+
+    function updateWithdrawAddress(
+        uint256 issuerId,
+        uint256 schemaHash,
+        address withdrawAddress
+    ) external ownerOrIssuer(issuerId, schemaHash) {
+        if (withdrawAddress == address(0)) {
+            revert InvalidWithdrawAddress('Invalid withdraw address');
+        }
+        PaymentData memory payData = getPaymentData(issuerId, schemaHash);
```

Comment:
More gas efficient to use storage pointer to the struct. Although that may be a minor problem here.

Reply:
replaced to storage